kafka-rest/src/main/java/io/confluent/kafkarest/controllers/SchemaManager.java

```diff
@@ -26,12 +27,42 @@
   /**
    * Returns the {@link RegisteredSchema schema} registered with the given {@code schemaId}.
    */
-  RegisteredSchema getSchemaById(EmbeddedFormat format, int schemaId);
+  RegisteredSchema getSchemaById(String subject, int schemaId);
 
   /**
-   * Parses and returns the potentially newly registered {@link RegisteredSchema schema}
-   * corresponding to {@code rawSchema}.
+   * Returns the {@link RegisteredSchema schema} registered with the given {@code schemaId}.
+   */
+  RegisteredSchema getSchemaById(
+      SubjectNameStrategy subjectNameStrategy, String topicName, boolean isKey, int schemaId);
+
+  /**
+   * Returns the {@link RegisteredSchema schema} registered with the given {@code schemaVersion}.
+   */
+  RegisteredSchema getSchemaByVersion(String subject, int schemaVersion);
+
+  /**
+   * Parses and returns {@code rawSchema}.
+   *
+   * <p>If the subject does not contain a registered schema equals to the parsed {@code rawSchema},
+   * one will be registered.
+   */
+  RegisteredSchema parseSchema(EmbeddedFormat format, String subject, String rawSchema);
+
+  /**
+   * Parses and returns {@code rawSchema}.
+   *
+   * <p>If the subject does not contain a registered schema equals to the parsed {@code rawSchema},
+   * one will be registered.
    */
   RegisteredSchema parseSchema(
-      EmbeddedFormat format, String topicName, String rawSchema, boolean isKey);
+      EmbeddedFormat format,
+      SubjectNameStrategy subjectNameStrategy,
```

Comment:
topic's the only valid value right? might be better to change the name of the method to be topic specific and remove this parameter. minor point though, internal detail.

Reply:
hmm.. naming is awkward. maybe good as is, though maybe `registerOrLookupSchema()`